File: frontend/src/core/types/knowledge.ts
export type RefreshFrequency = 'rarely' | 'standard' | 'often'

export interface KnowledgeLibraryDto {
  id: string
  name: string
  description: string | null
  nsfw: boolean
  document_count: number
  created_at: string
  updated_at: string
  default_refresh: RefreshFrequency
}

export interface KnowledgeDocumentDto {
  id: string
  library_id: string
  title: string
  media_type: 'text/markdown' | 'text/plain'
  size_bytes: number
  chunk_count: number
  embedding_status: 'pending' | 'processing' | 'completed' | 'failed'
  embedding_error: string | null
  created_at: string
  updated_at: string
  trigger_phrases: string[]
  refresh: RefreshFrequency | null
}

export interface KnowledgeDocumentDetailDto extends KnowledgeDocumentDto {
  content: string
}

export interface RetrievedChunkDto {
  library_name: string
  document_title: string
  heading_path: string[]
  preroll_text: string
  content: string
  score: number
}

File: shared/dtos/knowledge.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

RefreshFrequency = Literal["rarely", "standard", "often"]


class KnowledgeLibraryDto(BaseModel):
    id: str
    name: str
    description: str | None = None
    nsfw: bool = False
    document_count: int = 0
    created_at: datetime
    updated_at: datetime
    default_refresh: RefreshFrequency = "standard"


class KnowledgeDocumentDto(BaseModel):
    id: str
    library_id: str
    title: str
    media_type: Literal["text/markdown", "text/plain"]
    size_bytes: int
    chunk_count: int = 0
    embedding_status: Literal["pending", "processing", "completed", "failed"]
    embedding_error: str | None = None
    created_at: datetime
    updated_at: datetime
    trigger_phrases: list[str] = Field(default_factory=list)
    refresh: RefreshFrequency | None = None  # None = inherit from library


class KnowledgeDocumentDetailDto(KnowledgeDocumentDto):
    content: str


class CreateLibraryRequest(BaseModel):
    name: str
    description: str | None = None
    nsfw: bool = False


class UpdateLibraryRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    nsfw: bool | None = None


class CreateDocumentRequest(BaseModel):
    title: str
    content: str
    media_type: Literal["text/markdown", "text/plain"] = "text/markdown"


class UpdateDocumentRequest(BaseModel):
    title: str | None = None
    content: str | None = None
    media_type: Literal["text/markdown", "text/plain"] | None = None


class RetrievedChunkDto(BaseModel):
    library_name: str
    document_title: str
    heading_path: list[str]
    preroll_text: str
    content: str
    score: float


class SetKnowledgeLibrariesRequest(BaseModel):
    library_ids: list[str]
